rfcs/020 - Dependencies permissions RFC.md

```diff
@@ -0,0 +1,193 @@
+
+![Block at library level](./lib-lvl-v1.png)
+
+![Block at app level](./app-lvl-v1.png)
+
+- Define which external libraries will be part of the project and we could add specific blocks for dependecies used in the external library.
+
+The `@std/deps` will have the following API:
+
+```ts
+type Package {
+  std: Array<string>,
+  dependencies: Array<Dependency>,
+  block: Array<string>,
+  fullBlock: Array<string>,
+}
+
+type Dependency {
+  url: string, // Empty string or ignore it for global blocks at application level
+  block: Array<string>,
+  fullBlock: Array<string>,  // fullBlock, deepBlock, extensiveBlock, exhaustiveBlock?
+  package: Package,
+}
+
+event install: Package
+
+fn package(): Package {
+  return new Package {
+    std: [],
+    dependencies: [],
+    block: [],
+    fullBlock: [],
+  };
+}
+
+fn use(pkg: Package, std: [string]): Package {
+  pkg.std = std;
+  return pkg;
+}
+
+fn dependency(pkg: Package, url: string): Dependency {
+  return new Dependency {
+    url: url,
+    block: [],
+    fullBlock: [],
+    package: pkg,
+  };
+}
+
+fn add(dep: Depenency): Package {
+  const pkg = dep.pkg;
+  pkg.dependencies.push(dep);
+  return pkg;
+}
+
+fn block(pkg: Package, block: string): Package {
+  pkg.block.push(block);
+  return pkg;
+}
+
+fn block(pkg: Package, block: Array<string>): Package {
+  pkg.block = pkg.block + block;
+  return pkg;
+}
+
+fn block(dep: Dependency, block: string): Dependency {
+  dep.block.push(block);
+  return dep;
+}
+
+fn block(dep: Dependency, block: Array<string>): Dependency {
+  dep.block = dep.block + block;
+  return dep;
+}
+
+fn fullBlock(pkg: Package, block: string): Package {
+  pkg.fullBlock.push(block);
+  return pkg;
+}
+
+fn fullBlock(pkg: Package, block: Array<string>): Package {
+  pkg.fullBlock = pkg.fullBlock + block;
+  return pkg;
+}
+
+fn fullBlock(dep: Dependency, block: string): Dependency {
+  dep.fullBlock.push(fullBlock);
+  return dep;
+}
+
+fn fullBlock(dep: Dependency, block: Array<string>): Dependency {
+  dep.fullBlock = dep.fullBlock + block;
+  return dep;
+}
+
+fn commit(pkg: Package) {
+  // Remove std dep from blacklist
+  // Download and install each dep
+  // Apply blocks defined for each dependency
+  // Block/Enable (based on what we finally decide) depenencies at application level, meaning that mocks will exists at /dependencies/modules/
+}
+```
+
+The `.dependencies.ln` file could look something like:
+
+```ts
+on install fn (pkg: Package) = pkg
+  .use(['@std/tcp'])
+  // Block `@std/cmd` but do not override if any mock exists:
+  .dependency('https://github.com/org/new_dep_A')
+    .block('@std/cmd')
+    .add()
+  // Block `@std/cmd` and override mocks if any:
+  .dependency('https://github.com/org/new_dep_B')
+    .fullBlock('@std/cmd')
+    .add()
+  // Do not block any standrad library:
+  .dependency('https://github.com/org/new_dep_C')
+    .add()
+  .block('@std/fs')
+  .fullBlock(['@std/exec'])
```

Comment:
Can we include a comment above this section indicating that these are the explicitly blocked standard library modules for all dependencies? (Since `use` effectively defines the inverse of modules to insert blocking mocks.)